server/src/main/java/org/elasticsearch/discovery/HandshakingTransportAddressConnector.java

```diff
@@ -89,6 +90,13 @@ protected void doRun() throws Exception {
                     remoteNode = transportService.handshake(connection, probeHandshakeTimeout.millis());
                     // success means (amongst other things) that the cluster names match
                     logger.trace("[{}] handshake successful: {}", this, remoteNode);
+                } catch (Exception e) {
+                    // we opened a connection and successfully performed a low-level handshake, so we were definitely talking to an
+                    // Elasticsearch node, but the high-level handshake failed indicating some kind of mismatched configurations
+                    // (e.g. cluster name) that the user should address
+                    logger.warn(new ParameterizedMessage("handshake failed for [{}]", this), e);
```

Comment:
can we do this only for the handshake bit? Can we avoid logging if it is truly a connection failure? e.g. ConnectTransportException.

Reply:
I don't follow. As the comment says, we know that we successfully connected and performed a low-level handshake, so even a `ConnectTransportException` here seems worthy of a warning.

Reply:
I wanted to avoid for a flaky network connection to cause extra verbose warnings here. I'm fine if you want to leave it this way though